### ISSUE TITLE
How to average spinal cord segmentations across contrasts

### ISSUE THREAD

**Issue by sandrinebedard** [sct-pipeline/contrast-agnostic-softseg-spinalcord#7]
## Context
In order to generate a unique soft segmentation (related to #2), we want to average spinal cord segmentations across contrasts. However, because the images don't have the same FOV, we need to come up with a method on how to take this issue into account.

## Suggestion

* Put all voxels of images to 1.
* Apply transformation to align with T2w contrast. --> Use linear interpolation
* Add all warped images.
* Add all aligned segmentations. (also use linear interpolation for alignement)
* Divide the summed segmentation by the summed image (of 1).

**Comment by charleygros** [comment]
Your plan looks great @sandrinebedard !

We used a similar approach for the lesion mapping (see Eden et al 2018 and Kerbrat et al 2020), see these lines of code [here](https://github.com/neuropoly/lesion-mapping/blob/master/spinalcord/3_generate_LFM.py#L86).
Not straightforward to integrate into your pipeline as it is very much entangled w other stuffs. So, I'd recommend to do your own loop instead of pulling this project, quite simply:

When all your files are in the reference space (for you T2w):
- Sum all the SC seg files together: eg `sum_sc_seg.nii.gz`
- Sum all the coverage files together (eg images filled w ones indicating the original FOV): eg `sum_coverage.nii.gz`
- Divide `sum_sc_seg.nii.gz` by `sum_coverage.nii.gz`

**Comment by sandrinebedard** [comment]
Great! Thanks @charleygros!

**Comment by sandrinebedard** [comment]
So I've tested out the method! 

Two problems with the method:

* T1w SC segmentation and T2w don't necessarly start at the same slice, the value of the softseg can be of 0.5 at the top:

![image](https://user-images.githubusercontent.com/71230552/120711342-427f3d80-c48d-11eb-9914-c05473fab500.png)

* Some top and bottom slices segmentation are missing most of the time with `sct_deepseg_sc` on MTS data, so once again the averaging is affected. 
With manual correction, it can be fixed, and if so, it will definitely help with registration too!
Or we don’t count MTS data in the mean, but the registration is not that good at the end due to missing slices. 
![image](https://user-images.githubusercontent.com/71230552/120711970-06001180-c48e-11eb-96f8-1c2b5c8ad5c6.png)

I will need to explore with more data for MTS, but here is an example, even with manual correction, it will maybe be difficult to add the missing slices...
![image](https://user-images.githubusercontent.com/71230552/120712349-832b8680-c48e-11eb-8e3c-de64f7e4ceca.png)

**Comment by jcohenadad** [comment]
> To resolve this, maybe instead of putting all the images to 1, we could take the registered segmentation and put all non-zero values to 1, dilate the seg (we want the coverage map to include all the segs) and add those images toghether, so it would take into account only where we have the segmentation.

I re-read this several times but I still don't understand by "dilate the seg"? How should it be dilated, by how much, and how will that solve the "missing segmented slice" problem?

However, this seg dilation gave me an idea (maybe it is what you suggested): instead of creating the _coverage map_ by putting all-one to the native image, we could instead apply a _gigantic_ dilation (e.g. kernel of 500) to each slice of the segmentation. If there is a segmentation at a slice, then the whole slice will be ones. If there is no segmentation, the whole slice will be all zeros.

**Comment by sandrinebedard** [comment]
> However, this seg dilation gave me an idea (maybe it is what you suggested): instead of creating the _coverage map_ by putting all-one to the native image, we could instead apply a _gigantic_ dilation (e.g. kernel of 500) to each slice of the segmentation. If there is a segmentation at a slice, then the whole slice will be ones. If there is no segmentation, the whole slice will be all zeros.

Yes that was exactly the idea!!! Sorry if the explanation was not clear 🙂

**Comment by charleygros** [comment]
Note: oopsie, msg written 2hrs ago, forgot to send

> I will need to explore with more data for MTS, but here is an example, even with manual correction, it will maybe be difficult to add the missing slices...

Not sure to understand what you mean here. Can you help me.

> To resolve this, maybe instead of putting all the images to 1, we could take the registered segmentation and put all non-zero values to 1, dilate the seg (we want the coverage map to include all the segs) and add those images toghether, so it would take into account only where we have the segmentation

Question: Do you fill with ones before or after the resgistration? I think it should be done before registration --> at the top slices the MTS_reg (small IS coverage) will be zeros --> it will not contribute the the division of these slices.
In other words:
- slice 0: contrasts T1 and T2: division by 2
- slice 10: contrasts T1, T2 MTS: division by 3
- slice 58: contrasts: T1: division by one

Am I missing something here?

**Comment by sandrinebedard** [comment]
> > I will need to explore with more data for MTS, but here is an example, even with manual correction, it will maybe be difficult to add the missing slices...
> 
> Not sure to understand what you mean here. Can you help me.

I meant that the data quality is not always good for the first and last slices that we sometimes don't really see the spinal cord, so even if we wanted to manually correct those slices, it would not necessarily be acurate!

> What about doing a IS_masking (based on the sct_deepseg_sc output) on the coverage file before registration?
> Example: IS coverage: 0 to 100. IS coverage of seg: 10 to 98 --> put zeros to slices 0:10 and 99:100 on the coverage image
> 

yes that could be a solution too!

**Comment by sandrinebedard** [comment]
I've generated the coverage map using dilated segmentations, softseg results look great 😊
This softseg was generated using T1w, T2w, T2star, T1w_MT, MTon_MTS and dwi!
(btw, the surrounding dark red are values very close to zero)

![image](https://user-images.githubusercontent.com/71230552/120856339-b2ed9380-c54d-11eb-9bb0-d48e2b0f6b04.png)
